### PR TITLE
Add fake response testing

### PR DESCRIPTION
**Pull request by sztomi**
This is a continuation of #13 using [fake](https://github.com/cksac/fake-rs). This still doesn't test all possible response structures, but it tests a lot more. It might be useful to implement a form of fuzzing where the seeds can be recorded and similar tests can be ran. That shouldn't run in Github Actions though, because it will be potentially unstable.

### CODE REVIEW

File: dap/src/types.rs [hunk 1]
```diff
@@ -244,109 +258,151 @@ pub struct Capabilities {
   pub supports_single_thread_execution_requests: Option<bool>,
 }
 
+#[derive(Serialize, Deserialize, Debug, Default, Clone)]
+pub struct CustomValue(Value);
```
Comment (sztomi) [inline dap/src/types.rs:262]: This is added to be able to implement the `Dummy` trait for `Value`.

**Review by pileghoff — APPROVED** [review]
Overall it looks good, but i must admit, i haven't looked too deep into how all this works. Two things i noticed while playing around:
1. I wont detect if i remove `Option<_>` from a member (but it will detect if you add one, so thats good!). Not sure what can be done about that.
2. Because it does not test the entire message, but only the body, it did not catch that the raw string variant of `ResponseMessage` is wrong, as we need flatten it. Currently its wrapped in`Error(_)`. Im not sure how important it is to test the entire message in general, especially with all the issues you are encountering with the schema.

**Comment by sztomi** [comment]
> Because it does not test the entire message, but only the body, it did not catch that the raw string variant of ResponseMessage is wrong, as we need flatten it. 

Can you give me an example?

**Comment by pileghoff** [comment]
> > Because it does not test the entire message, but only the body, it did not catch that the raw string variant of ResponseMessage is wrong, as we need flatten it.
> 
> Can you give me an example?

```rust
fn main() {
  let a = Response {
    request_seq: 1,
    success: false,
    message: Some(ResponseMessage::Error("test".to_string())),
    body: None,
  };
  println!("{}", serde_json::to_string(&a).unwrap());
}
```

prints

```
{"request_seq":1,"success":false,"message":{"Error":"test"}}
```

but should print 
```
{"request_seq":1,"success":false,"message":"test"}
```

**Comment by pileghoff** [comment]
I also just now noticed that because we moved the dap source folder, `cargo run --examples` no longer works, as the examples folder also needs to be moved into the dap folder.

**Comment by sztomi** [comment]
Thanks, I filed an issue about the error flattening. It's a good point that we probably should test entire messages. That might be useful future work or we might ultimately end up just iterating enough (and adding individual regression tests) that we weed out all of these mistakes. 

I'll look into making the examples work again. I remember this being painful in workspaces.

**Comment by sztomi** [comment]
oh that somehow didn't get autosquashed when I rebased ... I'll amend on main.

**Pull request merged**